Publish blog post re: 2026 UI redesign (#20268)
pypa/warehouse: docs/blog/posts/2026-07-14-ui-updates.md

---

> title: "Changes are coming to the PyPI User Interface"

Reviewer: Feels a bit ominous, maybe:

- "Upcoming changes to the PyPI User Interface"
- "Preview of upcoming PyPI User Interface changes"
- "Planned updates to the PyPI User Interface"

?

description: We are rolling out updates to the PyPI user interface to improve security signaling and the display of package details.
authors:
  - nlh
date: 2026-07-14
tags:
  - user-interface
  - security
---

Over the next few months, we will be rolling out changes to the PyPI user interface, improving how we surface security signals and updating the pages where users view package details.

Updates will be staged to [TestPyPI](https://test.pypi.org/) and deployed to production in phases. This approach allows our team to thoroughly test the UI with production-like data, while providing the community with an opportunity to provide feedback.

> The first changes will be staged today.

Reviewer: ```suggestion
The first changes have already been staged today.
```

My thinking is, we should land the changes at testpypi before the blog post is announced, so users can click through and immediately see them after reading the blogpost.

<!-- more -->

> ## Context: Research and Style Guide

Reviewer: Maybe worth mentioning here that this UI hasn't changed significantly since it launched in 2018 and in the intervening years we have discovered some shortcomings (and what they are)? These changes aren't entirely driven by the OpenSSF effort.

Reviewer: I think between that and the following paragraphs something like this could help complete the history:

> In 2024, [PyPI introduced support for digital attestations](https://blog.pypi.org/posts/2024-11-14-pypi-now-supports-digital-attestations/), enabling PyPI package maintainers can now publish signed digital attestations when publishing, in order to further increase trust in the supply-chain security of their projects. This effort was just the first step, and notably did not include significant updates to PyPI to make it easier for users to understand and consume attestations.

> Between June and October 2025, the [OpenSSF Securing Repositories Working Group](https://repos.openssf.org/) partnered with [Superbloom](https://superbloom.design/) and [Kabu&nbsp;Creative](https://kabucreative.com/) to research and develop a [style guide](https://repos.openssf.org/attestations-style-guide) for effectively displaying and communicating attestations to software repository users, including PyPI users.

Reviewer: I think this doesn't make it clear that this involved interviews with real PyPI users, I think we should highlight that!

> The research identified that while PyPI currently supports the display of attestations, both the user interface and supporting documentation require significant updates. To address this, we are applying [level AAA (highest) recommendations from the style guide](https://repos.openssf.org/attestations-style-guide/level-aaa) to PyPI. This involves adding a new "Security" tab, while also taking the opportunity to improve security signaling for other PyPI metadata.

Reviewer: ```suggestion
Although PyPI currently supports the display of attestations, the research identified that both the user interface and supporting documentation require significant updates. To address this, we are applying [level AAA (highest) recommendations from the style guide](https://repos.openssf.org/attestations-style-guide/level-aaa) to PyPI. This involves adding a new "Security" tab, while also taking the opportunity to improve security signaling for other PyPI metadata.
```

## What’s Changing?
Key changes include:

> *   **Navigation moved to horizontal tabs:** We are moving internal navigation out of the sidebar and into a horizontal tab structure. This allows us to reserve the sidebar exclusively for metadata.

Reviewer: ```suggestion
*   **Navigation moved to horizontal tabs:** We are moving internal navigation out of the sidebar and into a horizontal tab structure. This allows us to reserve the sidebar exclusively for project- and release-related metadata.
```

*   **Metadata sidebar moved to the right:** Shifting the sidebar to the right places the most critical content—the package description/readme—on the left, aligning with a natural left-to-right reading pattern and mirroring the structure of other major platforms like GitHub and npm.

> *   **Separating Trust Levels:** We are explicitly distinguishing between official PyPI data, "Data verified by PyPI," and maintainer-provided content. This signaling will help users weigh data according to its source.

Reviewer: ```suggestion
*   **Separating Trust Levels:** We are explicitly distinguishing between project data intrinsic to PyPI (like when a release was created), data provided by a maintainer and verified by PyPI (like an upstream source repository) and maintainer-provided data that is unverified. This signaling will help users weigh the trustfulness of data according to its source.
```

*   **Optimized Sidebar Hierarchy:** Metadata is being reordered by utility, prioritizing project links and "freshness" signals (like release dates) at the top, while shifting classification data to the bottom to reduce cognitive load.

> *   **Dedicated Security Transparency:** We are introducing a new Security tab to house complex provenance and attestation metadata.

Reviewer: ```suggestion
*   **Dedicated Security Tab:** We are introducing a new Security tab to the project page in order to centralize data about a projects provenance and attestation metadata, as well as introducing clear indicators of when provenance is missing or has changed.
```

Also, I feel like this could be moved to the top of the list, this is the real highlight and focus of the funding effort!

*   **Phase 2: Files and Release History** (preview at [#20185](https://github.com/pypi/warehouse/issues/20185)) – Streamlining the file and release history tabs, adding attestation metadata where appropriate.
*   **Phase 3: Security Tab** (preview at [#20069](https://github.com/pypi/warehouse/issues/20069)) – Building a dedicated space for surfacing provenance and attestation data, surfacing warnings where appropriate.
*   **Phase 4: Documentation Overhaul** – Refreshing PyPI’s documentation to include security guidance and address documentation needs of different types of users.

## We Want Your Feedback
We invite you to test these changes on [TestPyPI](https://test.pypi.org/). 

If you have feedback, please share it on [this GitHub issue](https://github.com/pypi/warehouse/issues/20267).

If you find a bug or defect, please open a ticket on [our issue tracker](https://github.com/pypi/warehouse/issues). Please ensure you include browser and device details. This helps our team reproduce and fix the problem quickly.

## Acknowledgments

> We are grateful for the funding provided by the [Open&nbsp;Source&nbsp;Security&nbsp;Foundation&nbsp;(OSSF)](https://openssf.org/) for this initiative.

Reviewer: ```suggestion
We are grateful for the funding provided by the [Open&nbsp;Source&nbsp;Security&nbsp;Foundation&nbsp;(OSSF)](https://openssf.org/) for this initiative, as well as Dustin Ingram, Zach Steindler, and the members of the [OpenSSF Securing Software Repos WG](http://repos.openssf.org/) for proposing this project for funding and championing the larger effort to introduce attestations.
```

Feel free to reword, but Zach and I put a lot of work into making sure this could get funded!
